### DISCUSSION TITLE
Delay in building the project in dev mode

### DISCUSSION THREAD

**Discussion by geissondaSilva** [quarkusio/quarkus#37552] [Community]
I am migrating a legacy project that used thorntail as a framework.
We recently decided to migrate to quarkus to have better performance in production and development.
After the project configuration stage using only dependencies compatible with quarkus, we identified a construction time greater than that of thorntail. When analyzing the problem and performing profiling in the construction phase, we identified that a build stage was analyzing the project's beans to assemble the data for the quarkus devui page.

Class: io.quarkus.arc.deployment.devconsole.ArcDevConsoleProcessor.java
Quarkus version: 2.16.9-Final

The addNodesDependencies function of the ArcDevConsoleProcessor class was the one that spent the longest time running when starting the application (around 70% of the time consuming CPU)

We understand that the best way would be to break this project into several others. However, due to the urgency, we need to keep everything in one project only due to the high refactoring time.

My question is how can I disable this functionality to save build time in development mode in my project?

**Comment by maxandersen** [comment]
any chance you can try newer quarkus version? quarkus 3.4+ would potentially not have this calculation.

**Reply by geissondaSilva** [reply]
Due to the size of the project and the dependencies, it would be a very large effort. Ideally, it would be possible to create a configuration property to disable devui. Would it be possible to do this?

**Reply by maxandersen** [reply]
We don't currently have plans for doing more 2.x releases unless security issues and bug fixes. But let us see what @mkouba says what this would entail.

Have you tried make a patch that skips it ? if "just" in devui should be possible to try it.

**Reply by geissondaSilva** [reply]
I haven't gotten around to doing that yet. What I performed was a debugging process to find the cause of my problem. When performing an analysis using VisualVM, I identified that the "addNodesDependents" function consumes around 70% of the processing in the build phase.

If so, I may be patching this version.

**Reply by mkouba** [reply]
It should be straightforward to implement (I mean a config property to disable this Dev UI feature), although not "backportable" because the code has changed since 2.16. However, I suppose that with an app of this size (@geissondaSilva mentioned 4900 beans on zulip) you might expect other kind of problems too.

**Reply by geissondaSilva** [reply]
Because it is a legacy project, we are facing some problems. We are in the stage of migrating to quarkus at the moment. The next step would be to break this application into several services. In our approval, the only thing that is stopping us would be the build time in development mode. We have already generated a production version and have not yet had any reports of problems.

**Reply by mkouba** [reply]
I do understand that it's a legacy project ;-).

> We have already generated a production version and have not yet had any reports of problems.

That's great. If that's the case then we can try to add a config property to skip the generation of dependency graphs.

@phillip-kruger I wonder if it would make sense to add a config property to disable the Dev UI completely so that it would be possible to skip all the relevant build steps..?

**Reply by phillip-kruger** [reply]
Should be possible. This is for 2.x and maybe we add it to 3.x too?

**Reply by mkouba** [reply]
I think that we can do an ArC-specific config property in both and the Dev UI property in the main (3.7) only?

**Reply by mkouba** [reply]
@geissondaSilva It would be great if you could test the fix in this branch: https://github.com/mkouba/quarkus/tree/arc-dev-generate-dependency-graph-2-16; just use `quarkus.arc.dev-mode.generate-dependency-graphs=false` to skip the dependency graphs...

**Reply by geissondaSilva** [reply]
Perfect @mkouba, I'm going to start setting up the environment so I can test the functionality locally.

**Reply by gsmet** [reply]
@geissondaSilva https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md#checking-an-issue-is-fixed-in-main can help with that.

**Reply by geissondaSilva** [reply]
@mkouba I performed the local test and had a great result. my build time dropped from 5 minutes to 30 seconds.

Thanks a lot for the help. I thank the entire team.

**Reply by mkouba** [reply]
Ok, I will send a PR for the 2.6 branch and a separate one for the main branch. And maybe one for the 3.2 branch.

@gsmet Does it make sense? I mean it's a bit confusing since we're adding a config property that will be present in 2.6 but not in 3.0 etc.

**Reply by geissondaSilva** [reply]
@mkouba Is there any forecast for the release of the correction version?

**Reply by maxandersen** [reply]
@geissondaSilva just to set expectations. 2.x is not on a regular release schedule so not in any active plans. 

If this is urgent for you please use @mkoubas branch to build your own or if you have a red hat support you can open bug against red hat build of quarkus 2.13.

**Comment by mkouba** [comment]
This topic is also discussed on zulip: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/How.20to.20disable.20devui.20in.20dev.20mode.3F
